Dart/testSrc/com/jetbrains/dart/analysisServer/DartCallHierarchyTest.java

```diff
@@ -0,0 +1,186 @@
+            }
+            else {
+              return new DartCalleeTreeStructure(project, parent, HierarchyBrowserBaseEx.SCOPE_PROJECT);
+            }
+          }
+          else {
+            fail("Function not found");
+          }
+        }
+      }
+      return null;
+    }, fileNames);
+  }
+
+  PsiElement findReference(PsiElement dartClass, String referenceName) {
+    // Find any reference to the named function or method in the test code.
+    PsiElement[] result = new PsiElement[1];
+    try {
+      dartClass.acceptChildren(new DartRecursiveVisitor() {
+        @Override
+        public void visitReferenceExpression(@NotNull DartReferenceExpression reference) {
+          if (referenceName.equals(reference.getText())) {
+            if (reference.getNextSibling() == null) {
+              PsiElement parent = reference.getParent();
+              if (parent != null) {
+                IElementType type = parent.getNode().getElementType();
+                if (type == CALL_EXPRESSION) {
+                  result[0] = reference;
+                  throw new ExitVisitor();
+                }
+                parent = parent.getParent();
+                if (parent != null) {
+                  type = parent.getNode().getElementType();
+                  if (type == CALL_EXPRESSION) {
+                    List<PsiElement> results = new ArrayList<>();
+                    DartCallHierarchyTreeStructure.collectDeclarations(reference.resolve(), results);
+                    if (!results.isEmpty()) {
+                      result[0] = results.get(0);
+                      throw new ExitVisitor();
+                    }
+                  }
+                }
+              }
+            }
+          }
+          super.visitReferenceExpression(reference);
+        }
+      });
+    }
+    catch (ExitVisitor ex) {
+      return result[0]; // not null
+    }
+    return null;
+  }
+
+  public void testMethodCallers() throws Exception {
+    doCallHierarchyTest("B", "b", true, "B.dart", "A.dart");
+  }
+
+  public void testMethodCallees() throws Exception {
+    doCallHierarchyTest("B", "b", false, "B.dart", "C.dart");
+  }
+
+  public void testMethodRefCallers() throws Exception {
+    doCallHierarchyTest("B", "c", true, "B.dart", "A.dart", "C.dart");
+  }
+
+  public void testConstructorCallers() throws Exception {
+    doCallHierarchyTest("B", "b", true, "B.dart", "A.dart");
+  }
+
+  public void testConstructorCallees() throws Exception {
+    doCallHierarchyTest("B", "b", false, "B.dart", "A.dart", "C.dart");
+  }
+
+  public void testFunctionCallers() throws Exception {
+    doCallHierarchyTest(null, "b", true, "B.dart", "A.dart");
+  }
+
+  public void testFunctionCallees() throws Exception {
+    doCallHierarchyTest(null, "a", false, "A.dart", "C.dart", "B.dart");
+  }
+
+  public void testFunctionRefCallers() throws Exception {
```

Comment:
This test passes against SDK 1.15 but fails against 1.16. Can you please take a look?